docs/proposals/2170-kubeflow-training-v2/README.md

```diff
@@ -1692,6 +1692,61 @@ _Will be added after initial implementation for PyTorch._
 
 _Will be added after initial implementation for PyTorch._
 
+## Pipeline Framework
+
+We introduce the framework as internal mechanism so that we can easily expand mechanism
+for combination of Runtimes and TrainJob.
+
+The framework is called as Kubeflow Trainer Pipeline Framework, and it has 4 phases as you can see the following
+overview.
+
+![Overview](./TrainerPipelineFrameworkOverview.drawio.svg)
+
+As described in the following, each phase is basically executed step by step although `Startup Phase` is executed only once
+during starting trainer-controller-manager:
+
+- `Startup Phase`: Initialize internal components at once when the trainer-controller-manager starts.
+- `PreExecution Phase`: This phase is executed as a part of admission validating webhooks triggered by TrainJob is created and updated.
+- `Build Phase`: This phase is executed to build child Kubernetes resources and deploy those to the cluster.
+- `PostExecution Phase`: This phase is executed after the `Build Phase`.
+
+As you can see in the diagram, each phase has 2 types of APIs, `Internal API` and `Extension Point`.
+The Extension Point is exposed and could be added operations within the scope of the Pipeline Framework Plugins Interfaces as plugins
+and those plugins are performed in any order.
+On the other hand, the Internal APIs are not exposed and could not add any operations as opposed to the Extension Point.
+
+![Kubeflow TrainerPipelineFramework](./TrainerPipelineFramework.drawio.svg)
+
+- `Startup Phase`:
```

Comment:
I am wondering if we should also say that cluster operators can:
1. Register new plugins into existing TrainingRuntime and ClusterTrainingRuntime
2. Register new runtimes into runtime framework (e.g. SlurmRuntime).
Or you think, it is out-of-scope since we don't support registration of new runtimes yet ?

Reply:
> Register new runtimes into runtime framework (e.g. SlurmRuntime).
Or you think, it is out-of-scope since we don't support registration of new runtimes yet ?

Yeah, the current framework does not support inserting arbitrary plugins and runtimes (for sure, technically, we could do it). However, I think we can add another InternalAPI `Initialize Trainer Framework Pipelines` to diagram to clarify when pipelines are initialized. WDYT?

Reply:
Yeah, I think that would be nice.

Reply:
Done.